worker/codalabworker/aws_batch.py

```diff
@@ -529,6 +545,17 @@ def update(self):
 
 
 class Complete(AwsBatchRunState):
+    def recover(self, error):
+        # If the FSM failed because the batch job has gone missing, then report that and finish gracefully
+        self.update_metadata(run_status='Failed')
+        finalize_message = {
+            'exitcode': 500,
+            'failure_message': 'Could not complete Batch Job for unknown reason.\n\n%s' % str(error)
+        }
+        self._bundle_service.finalize_bundle(self._worker.id, self.uuid, finalize_message)
+        self._worker.finish_run(self.uuid)
+        return fsm.State.DONE
+
```

Comment:
I don't think it is mostly fetch_batch_job which fails. I actually suspect the most common failure is when codalab master requests timeout and then an uncaught exception causes the thread to die. Then if the thread ever gets restarted (which only happens if the worker is restarted as far as I can tell) and the job is not found it will hit this flow.

Reply:
If this code or my description indicated that `fetch_batch_job` was the root cause of our problems, then I must be being unclear somewhere because it is not. `fetch_batch_job` fails only when the batch job no longer exists which happens ~3 days after it is completed. This case happens when the FSM thread was killed a while ago and then the worker docker container is restarted. The aws_batch code will then load all the run states from the json file (including those whose thread has died) and start up their threads again. The ones which are too old will hit the case where the old code for `fetch_batch_job` fails because the job no longer exists.

Because I don't know the root cause of the failures, I wanted to put an upper bound on the number of times the Running state could fail. This is specifically in Running because it is relatively safe since it doesn't create/destroy resources, it only update the bundle metadata. As for the other states, they seemed much less safe to add recovery logic to without quite a bit of refactoring which I don't think is a good use of time right now.

I added the recovery logic to Complete mostly just to surface any errors so we don't need to dig through the very long and truncated logs to find them.

Reply:
Sorry, I was inferring that fetch_batch_job was the problem based on the fact you made a special exception for it. This is fine, thanks.